_posts/2023-06-28-#27307.md

```diff
@@ -7,18 +7,90 @@ authors: [ishaanam]
 ## Notes
+- In Bitcoin Core, every [wallet transaction](https://github.com/bitcoin/bitcoin/blob/7f0b79ea132d22ad5212c1d3ff4325715ca5ac12/src/wallet/transaction.h#L160)
+  has a [transaction state](https://github.com/bitcoin/bitcoin/blob/7f0b79ea132d22ad5212c1d3ff4325715ca5ac12/src/wallet/transaction.h#L68).
+  These transaction states are part of how the wallet decides which
+  transactions to allow the user to spend, and which transactions to
+  count towards a user's balance.
+
+- Wallet transaction states and conflicts were previously discussed in
+  review club [#27145](/27145)
+
+- The different Bitcoin Core transaction states are detailed
+  [here](https://gist.github.com/ishaanam/846adf3b453c3a85fe6e15c882c57ae0#types-of-tx-states).
+
+- When a transaction is considered `TxStateInactive`:
+    - Its inputs are considered spent
+    - Its outputs are not available to spend and don't count towards a
+      user's balance
+
+- When a transaction is considered `TxStateConflicted`:
+    - Its inputs are _not_ considered spent
+    - Its outputs are not available to spend and don't count towards a
+      user's balance
+
+- On master, transactions are only considered conflicted when a
+  transaction that spends one or more of the same transaction outputs is
+  mined into a block. This means that if a transaction is only
+  conflicted in the mempool, it is considered `TxStateInactive`. This
+  can can lead to [confusion](https://bitcoin-irc.chaincode.com/bitcoin-core-dev/2023-05-09#1683605157-1683612219;)
+  amongst users, because the funds seem to briefly "disappear".
+
+- This PR treats transactions with conflicts in the mempool as conflicted
+  as well, by adding another [transaction state for mempool-conflicted
+  transactions](https://github.com/bitcoin/bitcoin/pull/27307/files#diff-d41d68c5a65d67956c91b33ca86da7df1981d84a0b15b4a186deea566563fed5R46-R49)
+  and keeping track of the conflicting transactions in
+  [`MempoolConflicts`](https://github.com/bitcoin/bitcoin/pull/27307/files#diff-9ce137cd784ea308778842120aa2af6d2bb8369485b71f25e72b2a32cf0a5b21R316-R318),
+  a map of wallet tx hashes to a set of their mempool conflicts' tx
+  hashes.
+
+- This idea and a previous attempt to implement it is described
+  [here](https://github.com/bitcoin-core/bitcoin-devwiki/wiki/Wallet-Transaction-Conflict-Tracking#idea-mempool-conflicted)
 
 ## Questions
-1. Did you review the PR? [Concept ACK, approach ACK, tested ACK, or NACK](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review)? What was your review approach?
--->
+1. Did you review the PR? [Concept ACK, approach ACK, tested ACK, or
+NACK](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review)?
+What was your review approach?
+
+1. Is this PR fixing a bug or adding a feature? What is that
+bug/feature?
+
+1. What are the trade-offs with considering a mempool-conflicted
+transaction as conflicted instead of unconfirmed?
```

Comment:
"unconfirmed" does not seem to be a tx state? https://github.com/bitcoin/bitcoin/blob/7f0b79ea132d22ad5212c1d3ff4325715ca5ac12/src/wallet/transaction.h#L68

Reply:
Yes, I've changed it to "inactive".